File: backend/app/api/strategies.py
"""API endpoints for strategy center (策略中心)。

domain: requirement / selection / pricing / market
status: draft / testing / active / archived（只有 active 被业务引用）
"""
from fastapi import APIRouter, HTTPException
from app.repository.strategy_repo import StrategyRepository

router = APIRouter(prefix="/api/strategies", tags=["strategies"])

_VALID_STATUS = {"draft", "testing", "active", "archived"}
_VALID_DOMAIN = {"requirement", "selection", "pricing", "market"}


@router.get("/")
def list_strategies(domain: str = None, status: str = None, type: str = None):
    repo = StrategyRepository()
    try:
        return {"strategies": repo.list(domain=domain, status=status, type=type)}
    finally:
        repo.close()


@router.get("/{strategy_id}")
def get_strategy(strategy_id: int):
    repo = StrategyRepository()
    try:
        s = repo.get(strategy_id)
        if not s:
            raise HTTPException(404, f"Strategy {strategy_id} not found")
        return s
    finally:
        repo.close()


@router.post("/")
def create_strategy(data: dict):
    for k in ("domain", "type", "name"):
        if not data.get(k):
            raise HTTPException(400, f"Missing field: {k}")
    if data["domain"] not in _VALID_DOMAIN:
        raise HTTPException(400, f"Invalid domain, must be one of {sorted(_VALID_DOMAIN)}")
    repo = StrategyRepository()
    try:
        return repo.create(data, operator=data.get("operator", "system"))
    finally:
        repo.close()


@router.put("/{strategy_id}")
def update_strategy(strategy_id: int, data: dict):
    repo = StrategyRepository()
    try:
        s = repo.update(strategy_id, data, operator=data.get("operator", "system"))
        if not s:
            raise HTTPException(404, f"Strategy {strategy_id} not found")
        return s
    finally:
        repo.close()


@router.post("/{strategy_id}/status")
def set_status(strategy_id: int, data: dict):
    status = data.get("status")
    if status not in _VALID_STATUS:
        raise HTTPException(400, f"Invalid status, must be one of {sorted(_VALID_STATUS)}")
    repo = StrategyRepository()
    try:
        s = repo.set_status(strategy_id, status, operator=data.get("operator", "system"))
        if not s:
            raise HTTPException(404, f"Strategy {strategy_id} not found")
        return s
    finally:
        repo.close()


@router.delete("/{strategy_id}")
def delete_strategy(strategy_id: int):
    repo = StrategyRepository()
    try:
        if not repo.delete(strategy_id):
            raise HTTPException(404, f"Strategy {strategy_id} not found")
        return {"success": True}
    finally:
        repo.close()


@router.post("/{strategy_id}/usage")
def record_usage(strategy_id: int, data: dict):
    """引用埋点：记录某商机/报价引用了这条策略。"""
    repo = StrategyRepository()
    try:
        return repo.record_usage(
            strategy_id,
            version=data.get("version"),
            ref_type=data.get("ref_type"),
            ref_id=data.get("ref_id"),
            operator=data.get("operator", "system"),
        )
    finally:
        repo.close()


@router.get("/{strategy_id}/usage")
def usage_stats(strategy_id: int):
    repo = StrategyRepository()
    try:
        return repo.usage_stats(strategy_id)
    finally:
        repo.close()

File: frontend/src/api/strategies.ts
/**
 * 策略中心 API（对接 /api/strategies）。
 * domain: requirement/selection/pricing/market；status: draft/testing/active/archived
 */
import axios from 'axios'
import type { PolicyDocBody, StrategyModule } from '@/constants/policyMeta'

const RESP = <T>(p: Promise<{ data: T }>) => p.then(r => r.data)

export type StrategyDomain = 'requirement' | 'selection' | 'pricing' | 'market' | 'policy'
export type StrategyStatus = 'draft' | 'testing' | 'active' | 'archived'

export interface Strategy {
  id: number
  domain: StrategyDomain
  type: string
  name: string
  scope: Record<string, any> | null
  body: Record<string, any> | null
  status: StrategyStatus
  version: number
  change_reason: string | null
  description: string | null
  created_at: string | null
  updated_at: string | null
  created_by: string
  updated_by: string
}


export const strategyApi = {
  list: (params?: { domain?: StrategyDomain; status?: StrategyStatus; type?: string }) =>
    RESP<{ strategies: Strategy[] }>(axios.get('/api/strategies/', { params })),
  get: (id: number) => RESP<Strategy>(axios.get(`/api/strategies/${id}`)),
  create: (data: Partial<Strategy> & { domain: StrategyDomain; type: string; name: string }) =>
    RESP<Strategy>(axios.post('/api/strategies/', data)),
  update: (id: number, data: Partial<Strategy>) =>
    RESP<Strategy>(axios.put(`/api/strategies/${id}`, data)),
  setStatus: (id: number, status: StrategyStatus) =>
    RESP<Strategy>(axios.post(`/api/strategies/${id}/status`, { status })),
  delete: (id: number) => RESP<{ success: boolean }>(axios.delete(`/api/strategies/${id}`)),
  recordUsage: (id: number, data: { version?: number; ref_type?: string; ref_id?: string }) =>
    RESP<{ id: number }>(axios.post(`/api/strategies/${id}/usage`, data)),
  usageStats: (id: number) =>
    RESP<{ count: number; last_ref: string | null }>(axios.get(`/api/strategies/${id}/usage`)),
}

/**
 * 策略文档库（/api/policy-docs）—— 无数字 id，增删改查用「创建时间戳 (module, created_at)」定位。
 * 文档独立存 rules.policy_docs 表（2026-08-04 起，不再与策略规则混表、不再有自增 id）。
 */
export interface PolicyDoc {
  name: string
  module: StrategyModule
  category: string
  sort_order: number
  content_markdown: string
  description: string | null
  status: StrategyStatus
  version: number
  created_at: string | null      // 定位键（创建时间戳，不可变）
  updated_at: string | null
  created_by: string
  updated_by: string
  body: PolicyDocBody            // 兼容旧 readDocBody
}

export const policyDocApi = {
  list: (module?: StrategyModule, status?: StrategyStatus) =>
    RESP<{ docs: PolicyDoc[] }>(axios.get('/api/policy-docs/', { params: { module, status } })),
  create: (data: {
    module: StrategyModule; name: string; category: string; sort_order: number;
    content_markdown: string; description?: string; status?: StrategyStatus;
  }) =>
    RESP<PolicyDoc>(axios.post('/api/policy-docs/', data)),
  update: (data: {
    module: StrategyModule; created_at: string; name?: string; category?: string;
    sort_order?: number; content_markdown?: string; description?: string; status?: StrategyStatus;
  }) =>
    RESP<PolicyDoc>(axios.put('/api/policy-docs/', data)),
  remove: (module: StrategyModule, created_at: string) =>
    RESP<{ success: boolean }>(axios.delete('/api/policy-docs/', { params: { module, created_at } })),
}
